lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,398 @@
+    var baseLengthB = keyB.contains('*') ? keyB.indexOf('*') + 1 : keyB.length;
+    if (baseLengthA > baseLengthB) return -1;
+    if (baseLengthB > baseLengthA) return 1;
+    if (!keyA.contains("*")) return 1;
+    if (!keyB.contains("*")) return -1;
+    if (keyA.length > keyB.length) return -1;
+    if (keyB.length > keyA.length) return 1;
+    return 0;
+  }
+
+  /// Returns a file path to `subpath`, resolved in the `exports` object.
+  ///
+  /// Verifies the file exists relative to `packageRoot`. Instances of `*`
+  /// will be replaced with `patternMatch`.
+  ///
+  /// Implementation of `PACKAGE_TARGET_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _packageTargetResolve(
+      String? subpath, Object exports, String packageRoot,
+      [String? patternMatch]) {
+    switch (exports) {
+      case String string:
+        if (!string.startsWith('./')) {
+          throw "Export '$string' must be a path relative to the package root at '$packageRoot'.";
+        }
+        if (patternMatch != null) {
+          var replaced = string.replaceAll(RegExp(r'\*'), patternMatch);
+          var path = p.normalize(p.join(packageRoot, replaced));
+          return fileExists(path) ? path : null;
+        }
+        return p.join(packageRoot, string);
+      case Map<String, dynamic> map:
+        var conditions = ['sass', 'style', 'default'];
+        for (var (key, value) in map.pairs) {
+          if (!conditions.contains(key)) continue;
+          if (_packageTargetResolve(
+                  subpath, value as Object, packageRoot, patternMatch)
+              case var result?) {
+            return result;
+          }
+        }
+        return null;
+
+      case []:
+        return null;
+
+      case List<dynamic> array:
+        for (var value in array) {
+          var result = _packageTargetResolve(
+              subpath, value as Object, packageRoot, patternMatch);
+          if (result != null) {
+            return result;
+          }
+        }
+
+        return null;
+
+      default:
+        throw "Invalid 'exports' value in ${p.join(packageRoot, 'package.json')}";
+    }
+  }
+
+  /// Returns a path to a package's export without a subpath.
+  Object? _getMainExport(Object exports) {
+    Object? parseMap(Map<String, dynamic> map) {
+      if (!map.keys.any((key) => key.startsWith('.'))) {
+        return map;
+      }
+      if (map.containsKey('.')) {
+        return map['.'] as Object;
+      }
+      return null;
+    }
+
+    return switch (exports) {
+      String string => string,
+      List<String> list => list,
+      Map<String, dynamic> map => parseMap(map),
+      _ => null
+    };
+  }
+
+  /// Returns a list of all possible variations of `subpath` with extensions and
+  /// partials.
+  ///
+  /// `subpath` is part of the [bare import specifier], so we use the `/`
+  /// separator.
```

Comment:
"Using the `/` separator" doesn't mean anything. Separators are not the only _or_ the most relevant distinction between path syntaxes. A path is either a native filesystem path, in which case it should use `p.*`, or a URL, in which case it should use `p.url.*`. In this case, the `subpath`s being passed in are native paths, so this should use `p.*` everywhere.